src/NumericalAlgorithms/Spectral/Spectral.cpp

```diff
@@ -0,0 +1,312 @@
+  ASSERT(number_of_pts > 1,
+         "The number of collocation points must be at least two.");
+  ASSERT(number_of_pts <= ::Spectral::maximum_number_of_pts,
+         "Exceeded maximum number of collocation points.");
+  static const auto collo_pts_and_weights = []() {
+    std::vector<detail::CollocationPointsAndWeights<BasisType, QuadratureType>>
+        local_collo_pts;
+    local_collo_pts.reserve(::Spectral::maximum_number_of_pts - 1);
+    for (size_t n = 2; n <= Spectral::maximum_number_of_pts; ++n) {
+      local_collo_pts.emplace_back(n);
+    }
+    return local_collo_pts;
+  }();
+  return collo_pts_and_weights[number_of_pts - 2];
+}
+
+template <Basis BasisType, Quadrature QuadratureType>
+Matrix compute_differentiation_matrix(const size_t number_of_pts) noexcept {
+  // This implements algorithm 37 on p. 82 of Kopriva's book.
+  // It is valid for any collocation points and barycentric weights.
+  const DataVector& collocation_pts =
+      collocation_points_and_weights<BasisType, QuadratureType>(number_of_pts)
+          .collocation_pts();
+  const DataVector& bary_weights =
+      collocation_points_and_weights<BasisType, QuadratureType>(number_of_pts)
+          .barycentric_weights();
+
+  Matrix diff_matrix(number_of_pts, number_of_pts);
+  for (size_t i = 0; i < number_of_pts; ++i) {
+    double& diagonal = diff_matrix(i, i) = 0.0;
+    for (size_t j = 0; j < number_of_pts; ++j) {
+      if (LIKELY(i != j)) {
+        diff_matrix(i, j) =
+            bary_weights[j] /
+            (bary_weights[i] * (collocation_pts[i] - collocation_pts[j]));
+        diagonal -= diff_matrix(i, j);
+      }
+    }
+  }
+  return diff_matrix;
+}
+
+template <Basis BasisType, Quadrature QuadratureType>
+Matrix compute_spectral_to_grid_points_matrix(
+    const size_t number_of_pts) noexcept {
+  // Since u(x) = sum a_k P_k(x), matrix for u(x_j) is just P_k(x_j) transposed
+  const DataVector& collocation_pts =
+      collocation_points_and_weights<BasisType, QuadratureType>(number_of_pts)
+          .collocation_pts();
+  Matrix spec_to_grid(number_of_pts, number_of_pts);
+  // number_of_pts >=2 ASSERT'd elsewhere
+  for (size_t j = 0; j < number_of_pts; ++j) {
+    spec_to_grid(j, 0) = 1.0;  // P_0 = 1
+    double p_n_minus2 = 1.0;
+    double p_n_minus1 = collocation_pts[j];  // P_1 = x
+    spec_to_grid(j, 1) = p_n_minus1;
+    for (size_t k = 2; k < number_of_pts; ++k) {
+      const double p_n = ((2 * k - 1) * collocation_pts[j] * p_n_minus1 -
+                          (k - 1) * p_n_minus2) /
+                         k;
+      spec_to_grid(j, k) = p_n;
+      p_n_minus2 = p_n_minus1;
+      p_n_minus1 = p_n;
+    }
+  }
+  return spec_to_grid;
+}
+
+// GridPointsToSpectral is the inverse to the matrix above. Can be computed
+// numerically or with the analytic expressions P_k(x_j) w_j/gamma_k where
+// w_j=weight, gamma_k = normalization = 2/(2k+1), except = 2/N for k=N (for
+// LGL)
```

Comment:
this makes me suspicious that it is generic

Reply:
It probably is not. These functions were never tested and only ever used in the `linearize` function. I'd prefer doing this in a separate PR since it's low priority, but could make it part of this PR as well if you prefer (would delay things a bit).